homeassistant/components/volumio/browse_media.py

```diff
@@ -0,0 +1,157 @@
+    MEDIA_CLASS_DIRECTORY,
+    MEDIA_CLASS_GENRE,
+    MEDIA_CLASS_PLAYLIST,
+    MEDIA_CLASS_TRACK,
+    MEDIA_TYPE_MUSIC,
+)
+
+PLAYABLE_ITEM_TYPES = [
+    "folder",
+    "song",
+    "mywebradio",
+    "webradio",
+    "playlist",
+    "cuesong",
+    "remdisk",
+    "cuefile",
+    "folder-with-favourites",
+    "internal-folder",
+]
+
+NON_EXPANDABLE_ITEM_TYPES = [
+    "song",
+    "webradio",
+    "mywebradio",
+    "cuesong",
+    "album",
+    "artist",
+    "cd",
+    "play-playlist",
+]
+
+
+def _item_to_children_media_class(item, info=None):
+    if info and "album" in info and "artist" in info:
+        return MEDIA_CLASS_TRACK
+    if item["uri"].startswith("playlists"):
+        return MEDIA_CLASS_PLAYLIST
+    if item["uri"].startswith("artists://"):
+        if len(item["uri"]) > 10:
+            return MEDIA_CLASS_ALBUM
+        return MEDIA_CLASS_ARTIST
+    if item["uri"].startswith("albums://"):
+        if len(item["uri"]) > 9:
+            return MEDIA_CLASS_TRACK
+        return MEDIA_CLASS_ALBUM
+    if item["uri"].startswith("genres://"):
+        if len(item["uri"]) > 9:
+            return MEDIA_CLASS_ALBUM
+        return MEDIA_CLASS_GENRE
+    if item["uri"].startswith("Last_100") or item["uri"] == "favourites":
+        return MEDIA_CLASS_TRACK
+    if item["uri"].startswith("radio"):
+        return MEDIA_CLASS_CHANNEL
+    return MEDIA_CLASS_DIRECTORY
+
+
+def _item_to_media_class(item, parent_item=None):
+    if "type" not in item:
+        return MEDIA_CLASS_DIRECTORY
+    if item["type"] in ["webradio", "mywebradio"]:
+        return MEDIA_CLASS_CHANNEL
+    if item["type"] in ["song", "cuesong"]:
+        return MEDIA_CLASS_TRACK
+    if item.get("artist"):
+        return MEDIA_CLASS_ALBUM
+    if item["uri"].startswith("artists://") and len(item["uri"]) > 10:
+        return MEDIA_CLASS_ARTIST
+    if parent_item:
+        return _item_to_children_media_class(parent_item)
+    return MEDIA_CLASS_DIRECTORY
+
+
+def _list_payload(media_library, item, children=None):
+    return BrowseMedia(
+        title=item["name"],
+        media_class=MEDIA_CLASS_DIRECTORY,
+        children_media_class=_item_to_children_media_class(item),
+        media_content_type=MEDIA_TYPE_MUSIC,
+        media_content_id=json.dumps(item),
+        can_play=False,
+        can_expand=True,
+    )
+
+
+def _raw_item_payload(media_library, item, parent_item=None, title=None, info=None):
+    if "type" in item:
+        thumbnail = item.get("albumart")
+        if thumbnail:
+            thumbnail = media_library.canonic_url(thumbnail)
+    else:
+        # don't use the built-in volumio white-on-white icons
+        thumbnail = None
+
+    return {
+        "title": title or item.get("title"),
+        "media_class": _item_to_media_class(item, parent_item),
+        "children_media_class": _item_to_children_media_class(item, info),
+        "media_content_type": MEDIA_TYPE_MUSIC,
+        "media_content_id": json.dumps(item),
```

Comment:
Looks like you'll include more info than you need to identify the item?

Reply:
It's used when `play_media` is called. Volumio is expecting the item, not just its uri.

Reply:
And also forgot - also used to get things such as the thumbnail when navigation into the item.
I guess both can be done with an additional call to the server, but isn't it better to just store the required information?

Reply:
Just realized that the additional call to the server approach is more complicated - the way volumio browsing works, we'll also have to store the uri to the parent of the item, call browse on it, and find the item in the list.
I don't think there's a better approach than serializing the item to json.

Reply:
I doubt you have to serialize the value of `albumart` to be able to play it afterwards ?

Reply:
I need to serialize the url to the albumart (along with `type`), in order to correctly display the item when navigating it, because volumio's `browse` method gives you the full item information only when you browse the parent.
I probably need to serialize at least `type` and `service` to be able to play it (although their API examples show the client passing the full item, including album art).

Reply:
Can you try playing with it to see what's necessary? They have to look up the media anyway to play it, so they can fetch the album art etc then too ?

Reply:
Meh, okay.